src/libstore/build/derivation-goal.cc

```diff
@@ -311,14 +309,11 @@ void DerivationGoal::outputsSubstitutionTried()
     gaveUpOnSubstitution();
 }
 
+
 /* At least one of the output paths could not be
    produced using a substitute.  So we have to build instead. */
 void DerivationGoal::gaveUpOnSubstitution()
 {
-    /* Make sure checkPathValidity() from now on checks all
-       outputs. */
-    wantedOutputs.clear();
```

Comment:
I couldn't see any reason why `checkPathValidity()` needs to check all outputs. In any case, messing with `wantedOutputs` seems dirty...

Reply:
I agree it is dirty, but the idea I think (I did not write that but remember reading it IIRC) that since we are now doing the build, we will be making all afresh outputs, and we should fail if the derivation did not produce all the outputs even if we already had the missing outputs before.

Without this, we run the risk of some wonky non-deterministic derivation that produces a different subset of the outputs each time not failing as it should.

Reply:
Oh, another reason is that if during a build we discover we in fact need other outputs, it is nice to know we don't need to restart a build that got so upgraded. See `DerivationGoal::addWantedOutputs`

Reply:
A consequence of this is that I think `buildResultsWithPaths` is going to need to filter the results based on the derived path.

Reply:
From what I can tell, what outputs get build/registered (or whose absence causes the build to fail) is unaffected by `wantedOutputs` or `initialOutputs.*.wanted`.  So I think it should be okay. Restarts are only needed in the substitution case, not when we're actually building.

Reply:
I will fix this in the other PR too, sigh.

Reply:
Think what? I'm not sure I understand the problem...

Reply:
See the long PR description in https://github.com/NixOS/nix/pull/6312